### DISCUSSION TITLE
Log messages dispatched using the swift-log backend are missing normal swift-log formatting.

### DISCUSSION THREAD

**Discussion by thisisaaronland** [CocoaLumberjack/CocoaLumberjack#1389] [Q&A] [answered]
**Checklist**
- [x] I have read and understood the [CONTRIBUTING guide](https://github.com/CocoaLumberjack/CocoaLumberjack/blob/master/.github/CONTRIBUTING.md)
- [x] I have read the [Documentation](https://github.com/CocoaLumberjack/CocoaLumberjack#documentation)
- [x] I have searched for a similar issue in the [project](https://github.com/CocoaLumberjack/CocoaLumberjack/issues) and found none

**Describe the bug**

Log messages dispatched using the swift-log backend are missing normal swift-log formatting. Importantly, this might be an issue with our code because we are using CocoaLumberjack in a slightly unusual fashion.

Specifically, we are instantiating a local instance of `DDLogHandler.handlerFactory` (configured to use a `DDFileLogger` instance) which is then passed to the `Logger.MultiplexLogHandler` method:

https://github.com/sfomuseum/swift-sfomuseum-logger/blob/main/Sources/SFOMuseumLogger/SFOMuseumLogger.swift#L48-L64

And:

https://github.com/sfomuseum/swift-sfomuseum-logger/blob/main/Sources/SFOMuseumLogger/SFOMuseumLogger.swift#L78-L93

Messages are logged to the `DDFileLogger` instance but they are missing the default `swift-log` premable in addition to any metadata keys assigned to the logger. For example, the `StreamLogHandler.standardOutput` handler will emit:

```
2023-10-24T15:25:17-0700 info org.sfomuseum.logger : metadata=debug [sfomuseum_logger] hello world
```

But the `DDFileLogger` instance will only emit:

```
2023/10/24 22:25:17:955  hello world
```

**To Reproduce**

```
$> git clone https://github.com/sfomuseum/swift-sfomuseum-logger.git && cd swift-sfomuseum-logger
$> swift build && ./.build/debug/sfomuseum-logger --logfile true 'hello world'
```

https://github.com/sfomuseum/swift-sfomuseum-logger/blob/main/Scripts/sfomuseum-logger/SFOMuseumLogger.swift

**Expected behavior**

For logs written to the `DDFileLogger` instance to be formatted like this:

```
2023-10-24T15:25:17-0700 info org.sfomuseum.logger : metadata=debug [sfomuseum_logger] hello world
```

**Environment info**
  Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | macOS
 Platform Version        | 14.0
 CocoaLumberjack Version | 3.8.0
 Integration Method      | Swift Package Manager
 Xcode Version           | XCode 15.1
 Repro rate              | 100%
 Repro project link      | https://github.com/sfomuseum/swift-sfomuseum-logger

**Additional context**

As mentioned, I am open to the possibility that I am just passing incorrect, or incorrect, values to the local `cocoaLumberjackHandler` method (described above) but if that's the case I am not sure what I am doing wrong and any pointers would be appreciated.

Thanks,

**Comment by ffried** [comment, marked as answer]
@thisisaaronland First of all a big thanks to you for a very well written issue here! Including a reproduction project and a very detailed explanation. This helps a log to understand what you're trying to do and to help you out!

I've converted this to a discussion in our Q&A section, because you already (correctly) pointed out, that this might be an issue with your setup.

Now to the problem at hand:
This is actually working as designed. In `swift-log`, formatting is done by the log handler. What you see when using the `StreamLogHandler.standardOutput` is done [here](https://github.com/apple/swift-log/blob/main/Sources/Logging/Logging.swift#L1347). As soon as you use a different log handler (like the one CocoaLumberjack provides), this code is no longer executed.

You will need to do this formatting yourself using a `DDLogFormatter` and assign it to the `logFormatter` property on the `DDFileLogger` (or any `DDLogger` for that matter).

If you want consistent formatting, I'd actually recommend not using the `Logger.MultiplexLogHandler`, but using only CocoaLumberjacks LogHandler instead. Then you can use e.g. `DDOSLogger` for writing to `os_log` or use your own `DDLogger` implementation for writing to `stdout`. Both of the loggers can use the same log formatter and thus have the same formatting.

Small note, though: If you're using `DDOSLogger`, make your log formatter configurable to **not** include the date, because that will always be added by `os_log`.



Let me know if that helped or if you have any further questions!

**Comment by thisisaaronland** [comment]
Thanks for your reply. That was exactly what I needed to make things work the way I'd like.
